# Demand Forecasting & Inventory Risk Analytics

## Notebook 01 - Data Understanding

### Objective

'''Understand the dataset before performing SQL analysis, feature engineering, machine learning, and dashboard development.

---'''

In [5]:
import pandas as pd 
import numpy as np
print('imported')

imported


In [6]:
import os

print(os.getcwd())

C:\Users\ishik\OneDrive\Desktop\ishika\Supply-Chain-Analytics\notebooks


In [7]:
import os

print(os.listdir("../data"))

['beverage_distribution_data.csv']


In [8]:
df = pd.read_csv("../data/beverage_distribution_data.csv")
df.head()

,date,week_number,sku_id,sku_name,category,warehouse_id,warehouse_name,unit_price,lead_time_days,festival_flag,promo_flag,demand_units,available_stock_units,fulfilled_units,stockout_flag
0,2024-01-07,1,SKU01,Lager 650ml,Beer,WH01,North DC,120,5,0,0,1395,2254,1395,0
1,2024-01-14,2,SKU01,Lager 650ml,Beer,WH01,North DC,120,5,0,0,1674,3435,1674,0
2,2024-01-21,3,SKU01,Lager 650ml,Beer,WH01,North DC,120,5,0,1,2085,4337,2085,0
3,2024-01-28,4,SKU01,Lager 650ml,Beer,WH01,North DC,120,5,0,0,1812,2252,1812,0
4,2024-02-04,5,SKU01,Lager 650ml,Beer,WH01,North DC,120,5,0,1,1881,3016,1881,0


In [9]:
df.shape

(4992, 15)

'''this is a synthetic dataset which contains 4992 rows and 15 columns. 
Each row represents one SKU in 1 warehouse from one week'''


In [11]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4992 entries, 0 to 4991
Data columns (total 15 columns):
 #   Column                 Non-Null Count  Dtype 
---  ------                 --------------  ----- 
 0   date                   4992 non-null   object
 1   week_number            4992 non-null   int64 
 2   sku_id                 4992 non-null   object
 3   sku_name               4992 non-null   object
 4   category               4992 non-null   object
 5   warehouse_id           4992 non-null   object
 6   warehouse_name         4992 non-null   object
 7   unit_price             4992 non-null   int64 
 8   lead_time_days         4992 non-null   int64 
 9   festival_flag          4992 non-null   int64 
 10  promo_flag             4992 non-null   int64 
 11  demand_units           4992 non-null   int64 
 12  available_stock_units  4992 non-null   int64 
 13  fulfilled_units        4992 non-null   int64 
 14  stockout_flag          4992 non-null   int64 
dtypes: int64(9), object(6

## Observation

- Number of rows:4992
- Number of columns:15
- Numerical columns:9
- Categorical columns:6
- Missing values:0 '''

In [13]:
df.describe()

,week_number,unit_price,lead_time_days,festival_flag,promo_flag,demand_units,available_stock_units,fulfilled_units,stockout_flag
count,4992.000000,4992.000000,4992.000000,4992.000000,4992.000000,4992.000000,4992.000000,4992.000000,4992.000000
mean,52.500000,110.000000,6.166667,0.201923,0.085136,1115.850361,1610.128606,1066.015625,0.170473
std,30.023833,44.539092,1.818301,0.401475,0.279112,709.349920,964.636544,638.473500,0.376085
min,1.000000,30.000000,4.000000,0.000000,0.000000,179.000000,315.000000,179.000000,0.000000
25%,26.750000,85.000000,5.000000,0.000000,0.000000,569.750000,822.500000,546.750000,0.000000
50%,52.500000,115.000000,6.000000,0.000000,0.000000,958.000000,1436.000000,945.000000,0.000000
75%,78.250000,142.500000,6.500000,0.000000,0.000000,1475.000000,2180.500000,1439.000000,0.000000
max,104.000000,180.000000,10.000000,1.000000,1.000000,5688.000000,5094.000000,4372.000000,1.000000


''' ## Observation

The numerical summary helps understand the average demand, inventory levels and variability before building forecasting models.'''

In [15]:
df.isnull().sum()

date                     0
week_number              0
sku_id                   0
sku_name                 0
category                 0
warehouse_id             0
warehouse_name           0
unit_price               0
lead_time_days           0
festival_flag            0
promo_flag               0
demand_units             0
available_stock_units    0
fulfilled_units          0
stockout_flag            0
dtype: int64

In [16]:
df.nunique()

date                      104
week_number               104
sku_id                     12
sku_name                   12
category                    2
warehouse_id                4
warehouse_name              4
unit_price                 12
lead_time_days              6
festival_flag               2
promo_flag                  2
demand_units             2004
available_stock_units    2303
fulfilled_units          1827
stockout_flag               2
dtype: int64

In [17]:
df.columns

Index(['date', 'week_number', 'sku_id', 'sku_name', 'category', 'warehouse_id',
       'warehouse_name', 'unit_price', 'lead_time_days', 'festival_flag',
       'promo_flag', 'demand_units', 'available_stock_units',
       'fulfilled_units', 'stockout_flag'],
      dtype='object')

In [18]:
df.sample(5)

,date,week_number,sku_id,sku_name,category,warehouse_id,warehouse_name,unit_price,lead_time_days,festival_flag,promo_flag,demand_units,available_stock_units,fulfilled_units,stockout_flag
4699,2024-05-19,20,SKU10,Light Beer 330ml,Beer,WH04,East DC,90,6,0,0,993,1138,993,0
1396,2024-11-10,45,SKU02,Lager 330ml,Beer,WH02,West DC,70,5,1,0,3036,2912,2912,1
4588,2024-03-31,13,SKU09,Wheat Beer 650ml,Beer,WH04,East DC,160,9,0,0,297,420,297,0
4522,2024-12-22,51,SKU08,Lager Can 500ml,Beer,WH04,East DC,100,5,1,0,1281,1408,1281,0
781,2025-01-12,54,SKU08,Lager Can 500ml,Beer,WH01,North DC,100,5,0,0,1978,2231,1978,0


In [19]:
df.dtypes

date                     object
week_number               int64
sku_id                   object
sku_name                 object
category                 object
warehouse_id             object
warehouse_name           object
unit_price                int64
lead_time_days            int64
festival_flag             int64
promo_flag                int64
demand_units              int64
available_stock_units     int64
fulfilled_units           int64
stockout_flag             int64
dtype: object